### PR TITLE
Don't include implicit package references as dependencies in package when project is packed in most cases

### LINKED ISSUE #992: Avoid package downgrades for (implicit) NETStandard.Library package
Currently (in the 1.0 SDK), the implicit version of NETStandard.Library package is 1.6.1.  With the 2.0 tooling, this will change to (presumably) 2.0.0.  If someone creates a NuGet package targeting .NET Standard using the 2.0 tooling and someone else who is using the 1.0 tooling tries to reference that package from a .NET Standard project, they will get a package downgrade warning for NETStandard.Library.

We need to avoid this.

We could set `PrivateAssets="all"` on the implicit package reference in order to prevent it from being a dependency expressed on the NuGet packages that are generated.  That would prevent the package downgrade issue, but .NET Framework projects need the NETStandard.Library package to be referenced in order to get the right facades to support .NET Standard at runtime.

Ideally, the package generated would have a dependencies group for net45 (or whatever the lowest version of the .NET Framework supported would be) that would express a dependency on the NETStandard.Library package, which wouldn't apply to projects targeting other frameworks.  (It's possible that Xamarin or other targets might also need the dependency).

However, it's not possible to specify dependencies groups this way when creating a NuGet package from an SDK-style project.  There is a NuGet issue filed to enable this: https://github.com/NuGet/Home/issues/2300

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #992

Change metadata on implicit package references to NETStandard.Library and Microsoft.NETCore.App so that the package won't be a dependency of the package produced when packing the project in the following cases:

- The project is targeting .NET Standard 2.0 or higher
- The project is targeting .NET Core 2.0 or higher and the output type is not `Exe`

The implicit package reference will still be included when targeting 1.x so that updating to the 2.0 tools doesn't change the result of building or packing the project.  And it will still be included when targeting .NET Core and the output type is `Exe` as this is needed to make DotNetCliToolReferences work.

This PR also makes it explicit that even after there is a non-prerelease version of the NETStandard.Library 2.0 package, that projects targeting .NET Standard 1.x should still implicitly reference the 1.6.1 version of the package.  This is again so that the output of a project doesn't change when updating to the 2.0 tools.

I've also updated the build script to automatically create a binary log for the main SDK build.

Tests are still to come.

### CODE REVIEW

**Comment by dsplaisted** [comment]
I've finished adding the tests so this should be ready for review now @livarcocc @nguerrera

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePublishAssemblies.cs
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolvePublishAssemblies.cs]: How concerned are we about breaking changes to the public APIs of these Tasks?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets]: type-o `th`

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: copy-paste `NETStandard.Library`

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: This comment could use an update for the latest logic.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: Is this the right place to be updating these items?

**Comment by eerhardt** [comment]
>     Write-Host "  -Help                              Display this help message"

Would be nice to update this help text for the new parameter.

---
Refers to: build.ps1:28 in 7abe373. [](commit_id = 7abe373fa66ee4052f94ca5249d80cbed3e11a6a, deletion_comment = False)

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: The other option would be to put it in an `ItemGroup` in the .targets file but outside of any Target.  I thought that in this target was the right place for it to go, since this is the only place that metadata is consumed, and this keeps the behavior consistent if someone adds `PackageReference` items in a target, or after the import of the default .targets files.

**Comment by dsplaisted** [comment]
@ManishJayaswal @Pilchie for approval.  This is planned work for the 2.0 tools.  It avoids an issue that would otherwise occur when creating packages with the 2.0 tools and consuming them from the 1.0 tools.  See #992

**Comment by ManishJayaswal** [comment]
Approved pending 2 CRs and tests passing.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (rainersigwald) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: You don't need to specify `Update="@(PackageReference)"`--inside a target, update is the default operation. In fact, I'm very surprised this isn't erroring out for you--it should because of https://github.com/Microsoft/msbuild/pull/1620.

File: src/Tasks/Microsoft.NET.Build.Tasks/ResolvePublishAssemblies.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ResolvePublishAssemblies.cs]: I think semver is OK here (along similar lines that convinced me for Cli.Utils + the fact that there is no ship vehicle for consuming as a library). I believe there have been other breaks.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: I'm a little more concerned about breaks to "public" targets than public API.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: nit: separate lines for privateassets and publish

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: Comment doesn't match actual metadata used.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: Maybe both the itemgroup and the target that computes it should be "private" (underscore-prefixed) because adding to manipulating the item group directly doesn't give you anything more than setting the metadata, does it?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.props]: same nit about separate lines

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: I guess the most concerning thing is that the target is named `ComputeExcludeFromPublishPackageReferences`, and "computing" something doesn't necessary scream "modify the inputs" to me.

Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets]: @rainersigwald It looks like Microsoft/msbuild#1620 was never merged.  Did we ever end up making this an error?

**Comment by dsplaisted** [comment]
@dotnet-bot test Windows_NT_FullFramework Release

File: src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs
Comment (livarcocc) [inline src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: Why the == true?

Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs]: I think that's actually needed because the value you're comparing to is `Nullable<bool>`

File: test/Microsoft.NET.Pack.Tests/GivenThatThereAreImplicitPackageReferences.cs
Comment (livarcocc) [inline test/Microsoft.NET.Pack.Tests/GivenThatThereAreImplicitPackageReferences.cs]: There is a CoreMSBuildOnlyFactAttribute that you can use in all these cases. And there is a CoreMSBuildAndWindowsOnlyFactAttribute that you can use as well. Let's stick to this new approach now. Fix it here and in the other places you are adding as well.

**Review by livarcocc — APPROVED** [review]
Let's update the tests with the attribute before merging this PR to avoid creating this debt.

**Comment by dsplaisted** [comment]
@dotnet-bot test Windows_NT_FullFramework Release

**Pull request merged**
